fix resourceDashboardUpdate

dsbibby opened the pull request:
<!---
See what makes a good Pull Request at: https://hashicorp.github.io/terraform-provider-aws/raising-a-pull-request/
--->
### Description
<!---
Please provide a helpful description of what change this pull request will introduce.
--->
This PR addresses a bug in the aws_quicksight_dashboard resource.

When terraform updates a dashboard resource, two AWS API calls are made. The first updates the dashboard which results in a new dashboard version. The second call should update the published version of the dashboard to the new version.

The bug manifests when the first call takes a little time, which results in the second call returning an error about the dashboard creation still being in progress.

This is because the provider waits for the dashboard status to change to creation successful, however the API call used is checking the current published version, not the new version just made. As such it returns early and doesn’t actually wait.

This PR fixes this by explicitly waiting for the new version of the dashboard to be created before trying to update the published version.

Additionally this PR fixes two minor typo issues, and an issue caused when the state file for a template contains an empty string, resulting in "Error: a number is required" errors.

### Relations
<!---
If your pull request fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates.

For Example:

Relates #0000
or 
Closes #0000
--->
Closes #33463
Closes #35109

### References
<!---
Optionally, provide any helpful references that may help the reviewer(s).
--->
Redacted screenshot showing the failure:
![image-20230824-081431](https://github.com/hashicorp/terraform-provider-aws/assets/17053612/85e161d8-01b5-4d97-9adf-a8782e55f137)


### Output from Acceptance Testing
<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

Replace ec2 with the service package corresponding to your tests.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->

```console
% make testacc TESTS=TestAccQuickSightDashboard PKG=quicksight
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/quicksight/... -v -count 1 -parallel 20 -run='TestAccQuickSightDashboard'  -timeout 360m
?       github.com/hashicorp/terraform-provider-aws/internal/service/quicksight/schema  [no test files]
=== RUN   TestAccQuickSightDashboard_basic
=== PAUSE TestAccQuickSightDashboard_basic
=== RUN   TestAccQuickSightDashboard_disappears
=== PAUSE TestAccQuickSightDashboard_disappears
=== RUN   TestAccQuickSightDashboard_sourceEntity
=== PAUSE TestAccQuickSightDashboard_sourceEntity
=== RUN   TestAccQuickSightDashboard_updateVersionNumber
=== PAUSE TestAccQuickSightDashboard_updateVersionNumber
=== RUN   TestAccQuickSightDashboard_dashboardSpecificConfig
=== PAUSE TestAccQuickSightDashboard_dashboardSpecificConfig
=== CONT  TestAccQuickSightDashboard_basic
=== CONT  TestAccQuickSightDashboard_updateVersionNumber
=== CONT  TestAccQuickSightDashboard_sourceEntity
=== CONT  TestAccQuickSightDashboard_dashboardSpecificConfig
=== CONT  TestAccQuickSightDashboard_disappears
--- PASS: TestAccQuickSightDashboard_disappears (40.98s)
--- PASS: TestAccQuickSightDashboard_dashboardSpecificConfig (45.02s)
--- PASS: TestAccQuickSightDashboard_basic (45.43s)
--- PASS: TestAccQuickSightDashboard_sourceEntity (48.19s)
--- PASS: TestAccQuickSightDashboard_updateVersionNumber (95.68s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/quicksight 99.768s
```

dsbibby commented:
Hi, how do I encourage a review of this PR please? It's a pretty minor change to address a nasty bug that is stopping us (and surely others) from using this.

dev-slatto commented on internal/service/quicksight/dashboard.go:
```diff
@@ -203,7 +203,7 @@ func resourceDashboardRead(ctx context.Context, d *schema.ResourceData, meta int
 	}
 
-	out, err := FindDashboardByID(ctx, conn, d.Id())
+	out, err := FindDashboardByID(ctx, conn, d.Id(), -1)
```
Do we always want the latest version as this is hardcoded?

DavidB19456 replied:
That is the current (as in before this PR) behaviour, so no change to the current behaviour.

That said, we could easily add a version parameter to this `resourceDashboardRead()` function if that would be beneficial for the future?

atlesp replied:
The issue this MR tries to fix is urgent since the current provider version makes it impossible to update a dashboard. 

To update a dashboard we currently use the workaround to rename the dashboard, hence deleting the old and creating a new dashboard. This has the impact of making the user data, such as bookmarks, customisation, etc, vanish for the user and therefore not making it very user-friendly.

@dev-slatto I struggle to understand the concern you are raising. The old version of the code also tries to update to the latest version and it does not even allow you to update the dashboard at all since it now throws the error referred to in this MR.

When deploying a new version it would be (in my opinion) natural to deploy the latest version in most of the use cases. I struggle to see a good use case for not enabling the latest version except for maybe testing the new version before it is published to everyone. We have separate dev/test/prod env, so for us, it makes sense to always publish the latest.

dsbibby replied:
> Do we always want the latest version as this is hardcoded?

That is the current (as in before this PR) behaviour, so no change to the current behaviour.

That said, we could easily add a version parameter to this `resourceDashboardRead()` function if that would be beneficial for the future?

dev-slatto replied:
Had a chat with Graham and we concluded that it makes sence for a hardcoded version for the resource, and that a data source of dashboard should be able to include the version paramater. However, see the comment later down in the PR about defining the version as a constant instead of hardcoding it 😄

dsbibby replied:
Great, I have updated to use a constant 👍

gdavison commented on internal/service/quicksight/dashboard_test.go:
```diff
@@ -156,6 +156,8 @@ func TestAccQuickSightDashboard_updateVersionNumber(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "name", rNameUpdated),
 					resource.TestCheckResourceAttr(resourceName, "status", quicksight.ResourceStatusCreationSuccessful),
 					resource.TestCheckResourceAttr(resourceName, "version_number", "2"),
+					testAccCheckDashboardVersionExists(ctx, resourceName, 1, &dashboard),
+					testAccCheckDashboardName(&dashboard, rName),
```
For clarity, these should indicate that they're referencing version 1, e.g.

```suggestion
					testAccCheckDashboardVersionExists(ctx, resourceName, 1, &dashboardV1),
					testAccCheckDashboardName(&dashboardV1, rName),
```

dsbibby replied:
Updated to `dashboardV1`

gdavison reviewed (changes requested):
Thanks for the PR, @dsbibby. It looks good, though I have a few suggestions

DavidB19456 commented on internal/service/quicksight/dashboard.go:
```diff
@@ -297,14 +297,15 @@ func resourceDashboardUpdate(ctx context.Context, d *schema.ResourceData, meta i
 		}
 
-		if _, err := waitDashboardUpdated(ctx, conn, d.Id(), d.Timeout(schema.TimeoutUpdate)); err != nil {
+		updatedVersionNumber := extractVersionFromARN(aws.StringValue(out.VersionArn))
```
Have addressed this 👍

dsbibby replied:
Have addressed this 👍

dsbibby commented on internal/service/quicksight/dashboard.go:
```diff
@@ -367,15 +368,26 @@ func resourceDashboardDelete(ctx context.Context, d *schema.ResourceData, meta i
 }
 
-func FindDashboardByID(ctx context.Context, conn *quicksight.QuickSight, id string) (*quicksight.Dashboard, error) {
+// Pass version as -1 for latest published version, or specify a specific version if required
```
Done!

dsbibby commented on internal/service/quicksight/dashboard.go:
```diff
@@ -367,15 +368,26 @@ func resourceDashboardDelete(ctx context.Context, d *schema.ResourceData, meta i
+			DashboardId:   aws.String(dashboardId),
+			VersionNumber: &version,
+		}
 	}
```
Yes, that's neater, done 👍

dsbibby commented:
I have addressed the comments so far. In testing this with a massive config I have access to, I got blocked by a couple of minor typos for which there are existing issues and a regression introduced by #33931 - I added fixes for these also so that I could complete my tests. Hope that is OK.

gdavison reviewed (approved):
Looks good! 🚀 

--- PASS: TestAccQuickSightDashboard_disappears (31.92s)
--- PASS: TestAccQuickSightDashboard_dashboardSpecificConfig (32.84s)
--- PASS: TestAccQuickSightDashboard_sourceEntity (35.41s)
--- PASS: TestAccQuickSightDashboard_basic (35.93s)
--- PASS: TestAccQuickSightDashboard_updateVersionNumber (57.83s)
